### ISSUE TITLE
Shadow glitch

### ISSUE THREAD

**Issue by k8yb8t** [pygfx/pygfx#476]
Hello everyone, been playing with your examples. Remarkable code; god level commenting; thank you.
Looking at your 'Types of Cylinders' example [https://pygfx.readthedocs.io/en/latest/_gallery/feature_demo/cylinder.html#sphx-glr-gallery-feature-demo-cylinder-py](url), I'm convinced the shadows cast are incorrect. The full cylinder seems okay, but the other two objects appear to be missing shadows that ought to be cast by the faces closest the light source.
Don't know code well enough to contribute there I'm afraid; just trying to help :)

**Comment by Korijn** [comment]
Thank you for the kind words!

Could be a front/back face culling problem perhaps?
@panxinmiao what do you think?

**Comment by almarklein** [comment]
Hey! Thanks for reporting this. I just ran the example and examined the shadows, but I don't see anything off, to be honest. If I move the camera so that the object occludes the shadow, it's a perfect fit.

**Comment by k8yb8t** [comment]
Okay, got some screen grabs for you, which should at least explain how I'm looking at the scene.

In 1 & 2 I'm trying to show that the upper half of the open cylinder, (y==0 to y==height/2 ?), doesn't appear to be represented in the shadow, i.e. the slight overhang to the right in this perspective.
![shadows-1](https://user-images.githubusercontent.com/127444420/224259535-2b7c3efa-5051-4068-aef6-a854d9728729.png)
![shadows-2](https://user-images.githubusercontent.com/127444420/224259582-84e7dd32-9c0d-4cc5-ac03-1736e98f2c75.png)

In 3 & 4 the shadow has a 'dint' in either side, which seems consistent with the lower half of the object, but the closest face to the light source has parallel edges; would that not occlude what we see?

![shadows-3](https://user-images.githubusercontent.com/127444420/224259646-a004ac09-15fe-43c8-8d05-3e155bfbade9.png)
![shadows-4](https://user-images.githubusercontent.com/127444420/224259679-d09550d9-88d8-4577-84bf-b8082b401a8f.png)

Forgive me if I'm barking up the wrong tree here...

**Comment by Korijn** [comment]
You can change the example like this to see the position of the point light:

```diff
diff --git a/examples/feature_demo/cylinder.py b/examples/feature_demo/cylinder.py
index e1d99b6..2de8165 100644
--- a/examples/feature_demo/cylinder.py
+++ b/examples/feature_demo/cylinder.py
@@ -59,6 +59,7 @@ controller.add_default_event_handlers(renderer, camera)

 scene.add(gfx.AmbientLight())
 light = gfx.PointLight(position=(0, 70, 0))
+light.add(gfx.PointLightHelper())
 light.cast_shadow = True
 scene.add(light)
```

That makes it a little easier to position the camera correctly:

![image](https://user-images.githubusercontent.com/1882046/224278016-270c8fad-247e-4b43-9fb3-87a193df86dc.png)
![image](https://user-images.githubusercontent.com/1882046/224278051-7c8a4d84-ca55-4a83-88f9-d656fd0aa08f.png)
![image](https://user-images.githubusercontent.com/1882046/224278179-25b712a7-6abb-4a2a-a4a8-8ee2d3225da6.png)
![image](https://user-images.githubusercontent.com/1882046/224278229-5825bbee-6e48-480c-b9fe-0d67899d0536.png)

And so it seems you are right!

**Comment by k8yb8t** [comment]
If the same thing is happening with the green cylinder, I'm guessing the hidden end-cap is generating the 'complete' shadow at the top?

**Comment by almarklein** [comment]
I guess I did not look close enough. Great to have this resolved!

> If the same thing is happening with the green cylinder, I'm guessing the hidden end-cap is generating the 'complete' shadow at the top?

Yeah, that's very likely

> And this actually also fixes the dots of light on the ground we could see in https://github.com/pygfx/pygfx/issues/453 🎉

Mystery solved! 🥳

**Comment by k8yb8t** [comment]
At the risk of losing friends and alienating everyone... if the light sources in this scene, (re: the fires in the bowls), are actually 'in' the bowls, then the shadows appear flawed in what might be a similar manner.
![image](https://user-images.githubusercontent.com/127444420/224295320-c3e74a66-b520-434c-a941-6dac018d6540.png)
Honestly people, no disrespect implied or intended ;)

**Comment by Korijn** [comment]
They're point lights floating a little bit above the bowls. I didn't feel like putting nice flame sprites in the bowls, but you can imagine there's a little flame dancing around in them. :)

**Comment by panxinmiao** [comment]
Ah, I just saw this discussion. This is a long story. 😅 

"cull_mode" is intentionally set to "front" to eliminate "shadow acne", but it should only be correct for objects with "volume".

In general, instead of using "front culling", you can use appropriate "shadow bias" to eliminate ”shaodow acne“ (our shadow API actually provides this ability),  but this will bring "peter panning".

When you do not use "front culling" nor set a ”shadow bias“, "shadow acne" will appear on some surfaces (in short, "shadow acne" can only occur when a surface can be seen by our eyes and light at the same time).

So, in order to achieve good shadow effect, you should carefully set which objects in the scene will cast shadows,  and the wall in the sponza scene no need to cast shadows (because no objects in the scene will receive their shadows).  When it does not cast shadows, the "shadow acne" on the wall will disappear, and we don't have to be in a dilemma about whether to use "front culling" or set a "shadow bias".

For this scenario, setting a appropriate small ”shadow bias“ instead "front culling" (without obvious "peter panning") maybe achieve better results.

In short, different schemes are suitable for different situations. Generally speaking,When the objects casting shadows in the scene are all entities with volume (not a surface), it is the simplest and most effective to use "front culling".

But I do think that we should provide a way for users to choose whether to use "front culling" when an object casts shadows.

I have some trivial matters in my work recently. I will think about it again next week.

**Comment by Korijn** [comment]
Wow, so many new terms!

> Shadow acne (Figure 8), a term synonymous with erroneous self-shadowing, occurs when the shadow map quantizes the depth over an entire texel. When the shader compares an actual depth against this value, it is as likely to be self-shadowed as it is to be unshadowed.

That explains why it looks like z fighting. Interesting!

**Comment by almarklein** [comment]
What if Peter pan got shadow acne? 😆

**Comment by almarklein** [comment]
> This article explains shadow acne and peter panning very well, with some helpful illustrations:
> 
> https://learnopengl.com/Advanced-Lighting/Shadows/Shadow-Mapping

Nice read indeed!
